libtenzir/builtins/formats/leef.cpp

```diff
@@ -137,40 +128,56 @@ auto parse_delimiter(std::string_view field) -> std::variant<char, diagnostic> {
   return field[0];
 }
 
-// Converts a raw, unescaped string to a data instance.
-auto to_data(std::string_view str) -> data {
-  auto unescaped = unescape(str);
-  auto result = data{};
-  if (not(parsers::data - parsers::pattern)(unescaped, result)) {
-    result = std::move(unescaped);
-  }
-  return result;
-};
-
 /// Parses the LEEF attributes field as a sequence of key-value pairs.
-auto parse_attributes(char delimiter, std::string_view attributes)
-  -> std::variant<record, diagnostic> {
-  const auto key = +(parsers::printable - '=');
-  const auto value = *(parsers::printable - delimiter);
-  const auto kvp = key >> '=' >> value;
-  const auto kvp_list = kvp % delimiter;
-  auto kvps = std::vector<std::pair<std::string, std::string>>{};
-  if (not kvp_list(attributes, kvps)) {
-    return diagnostic::warning("failed to parse LEEF attributes")
-      .note("attributes: {}", attributes)
-      .done();
-  }
-  auto result = record{};
-  result.reserve(kvps.size());
-  for (auto& [key, value] : kvps) {
-    result.emplace(std::move(key), to_data(value));
+auto parse_attributes(char delimiter, std::string_view attributes,
+                      auto builder) -> std::optional<diagnostic> {
+  while (not attributes.empty()) {
+    auto attr_end = attributes.find(delimiter);
+    auto sep_pos = attributes.find('=');
+    if (sep_pos == 0) {
+      return diagnostic::warning(
+               "Ill-formed event missing key before separator")
+        .note("attribute was `{}`", attributes.substr(0, attr_end))
+        .done();
+    }
+    while (attributes[sep_pos - 1] == '\\' and sep_pos > 1
+           and attributes[sep_pos - 2] != '\\') {
+      sep_pos = attributes.find('=', sep_pos + 1);
+    }
+    if (sep_pos == attributes.npos) {
+      return diagnostic::warning("Ill-formed event missing "
+                                 "key-value separator in attribute")
+        .note("attribute was `{}`", attributes.substr(0, attr_end))
+        .done();
+    }
+    auto key = attributes.substr(0, sep_pos);
+    auto value = attributes.substr(sep_pos + 1, attr_end - sep_pos - 1);
```

Comment:
This might coincidentally work, but looks dangerous for `attr_end == npos`. You probably want to add some check here, or alternatively: The whole thing should become a lot simpler if you first extract a `string_view` for the `attribute`, and then use that to `substr` here.